src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs

```diff
@@ -114,204 +157,53 @@ public override void Initialize(AnalysisContext analysisContext)
                     int actualArgumentCount = intializer.ElementValues.Length;
                     if (actualArgumentCount != expectedStringFormatArgumentCount)
                     {
-                        operationContext.ReportDiagnostic(operationContext.Operation.Syntax.CreateDiagnostic(Rule));
+                        context.ReportDiagnostic(invocation.CreateDiagnostic(IncorrectNumberOfArgsRule));
                     }
                 }, OperationKind.Invocation);
             });
         }
 
-        private static int GetFormattingArguments(string format)
+        private static HashSet<int> GetStringFormatItemIndexes(string stringFormat)
         {
-            // code is from mscorlib
-            // https://github.com/dotnet/coreclr/blob/bc146608854d1db9cdbcc0b08029a87754e12b49/src/mscorlib/src/System/Text/StringBuilder.cs#L1312
+            // removing escaped left brackets and replacing with space characters so they won't
+            // impede the extraction of placeholders, yet the locations of the placeholders are
+            // the same as in the original string.
+            var formatStringWithEscapedBracketsChangedToSpaces = s_removeEscapedBracketsRegex.Replace(stringFormat, "  ");
 
-            // return count of this format - {index[,alignment][:formatString]}
-            var pos = 0;
-            int len = format.Length;
-            var uniqueNumbers = new HashSet<int>();
+            var matches = s_extractPlaceholdersRegex.Matches(formatStringWithEscapedBracketsChangedToSpaces);
+            var indexes = new HashSet<int>();
 
-            // main loop
-            while (true)
+            foreach (Match match in matches)
```

Comment:
Any preference between the explicit foreach and a `Select` call? Note that I will need to do a cast first.

Reply:
I think foreach auto-casting is fine